### PR TITLE
o/devicestate: validate volume info from installer

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Validate the volume information provided to the installer API against
the gadget from the system label to install.

### CODE REVIEW

File: overlord/devicestate/handlers_install.go [hunk 1]
```diff
@@ -1284,6 +1286,174 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
 	return sys, snapInfos, snapSeeds, mntPtForType, unmount, nil
 }
 
+func structureFromName(name string, Structures []gadget.VolumeStructure) (*gadget.VolumeStructure, error) {
+	var foundStruct *gadget.VolumeStructure
+	for _, structure := range Structures {
+		// Future go versions will have per-iter variables... but not 1.13
+		structure := structure
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:1293]: This is a bit too subtle IMHO, I would keep the comment and expand it to explain how the code should change after we require a more recent go version, but meanwhile use a different variable name here.

File: overlord/devicestate/handlers_install.go [hunk 2]
```diff
@@ -1284,6 +1286,174 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
 	return sys, snapInfos, snapSeeds, mntPtForType, unmount, nil
 }
 
+func structureFromName(name string, Structures []gadget.VolumeStructure) (*gadget.VolumeStructure, error) {
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:1289]: I'm used to search functions having the collection as first parameter, and the search query as the second one (cfr. `strstr()`).

Comment (mardy) [reply overlord/devicestate/handlers_install.go:1289]: Another nitpick is `Structures` with the capital `S`, is there a reason for it?

File: overlord/devicestate/handlers_install.go [hunk 3]
```diff
@@ -1284,6 +1286,174 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
...
+	return foundStruct, nil
+}
+
+func compareOffsets(offset1, offset2 *quantity.Offset) error {
+	if (offset1 == nil && offset2 != nil) || (offset1 != nil && offset2 == nil) {
+		return fmt.Errorf("cannot compare offsets, one is nil")
+	}
+	if offset1 != nil && offset2 != nil {
+		if *offset1 != *offset2 {
+			return fmt.Errorf("offset %d does not match %d from gadget.yaml", *offset1, *offset2)
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:1315]: nitpick (also for `compareOffsetWrite` below): if we assume that `offset2` is the one from `gadget.yaml`, I'd rename it to `offsetGadget` or just `gadget`. Or maybe add an additional parameter which is a string that describes where the second offset comes from, and then you build the error message using this information?

I'm probably overthinking it, but it feels a bit strange to have a function with such a generic name that emits error messages specific to the caller's implementation.

File: overlord/devicestate/handlers_install.go [hunk 4]
```diff
@@ -1284,6 +1286,174 @@ func (m *DeviceManager) loadAndMountSystemLabelSnaps(systemLabel string) (
...
+	isFile := true
+	if contentPath == "" {
+		contentPath = content.Image
+		isFile = false
+	}
+
+	var foundCont *gadget.VolumeContent
+	for _, gadgetCont := range gadgetContents {
+		// Future go versions will have per-iter variables... but not 1.13
+		gadgetCont := gadgetCont
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:1346]: same here

**Review by mardy — COMMENTED** [review]
Mostly nitpicks.

**Comment by alfonsosanchezbeato** [comment]
Closing as things have already changed quite a bit and the validation will reuse code from the gadget package.

**Pull request closed without merging**